specs/sharding/beacon-chain.md

```diff
@@ -103,786 +87,558 @@ The following values are (non-configurable) constants used throughout the specif
+#### `eval_poly_at`
 
 ```python
-def get_active_shard_count(state: BeaconState, epoch: Epoch) -> uint64:
+def eval_poly_at(poly: List[BLSFieldElement], x: BLSFieldElement) -> BLSFieldElement:
     """
-    Return the number of active shards.
-    Note that this puts an upper bound on the number of committees per slot.
+    Evaluates a polynomial (in evaluation form) at an arbitrary point
     """
-    return INITIAL_ACTIVE_SHARDS
+    field_elements_per_blob = SAMPLES_PER_BLOB * FIELD_ELEMENTS_PER_SAMPLE
+    roots = roots_of_unity(field_elements_per_blob)
+
+    def A(z):
+        r = 1
+        for w in roots:
+            r = r * (z - w) % MODULUS
+        return r
+
+    def Aprime(z):
+        return field_elements_per_blob * pow(z, field_elements_per_blob - 1, MODULUS) 
+
+    r = 0
+    inverses = [modular_inverse(z - x) for z in roots]
+    for i, x in enumerate(inverses):
+        r += f[i] * modular_inverse(Aprime(roots[i])) * x % self.MODULUS
+    r = r * A(x) % self.MODULUS
+    return r
 ```
 
-#### `get_shard_proposer_index`
+#### `next_power_of_two`
 
 ```python
-def get_shard_proposer_index(state: BeaconState, slot: Slot, shard: Shard) -> ValidatorIndex:
-    """
-    Return the proposer's index of shard block at ``slot``.
-    """
-    epoch = compute_epoch_at_slot(slot)
-    seed = hash(get_seed(state, epoch, DOMAIN_SHARD_BLOB) + uint_to_bytes(slot) + uint_to_bytes(shard))
-    indices = get_active_validator_indices(state, epoch)
-    return compute_proposer_index(state, indices, seed)
+def next_power_of_two(x: int) -> int:
+    return 2 ** ((x - 1).bit_length())
 ```
 
-#### `get_start_shard`
+#### `low_degree_check`
 
 ```python
-def get_start_shard(state: BeaconState, slot: Slot) -> Shard:
+def low_degree_check(commitments: List[KZGCommitment]):
+    """
+    Checks that the commitments are on a low-degree polynomial
+    If there are 2*N commitments, that means they should lie on a polynomial
+    of degree d = K - N - 1, where K = next_power_of_two(2*N)
+    (The remaining positions are filled with 0, this is to make FFTs usable)
+
+    For details see here: https://notes.ethereum.org/@dankrad/barycentric_low_degree_check
+    """
+    assert len(commitments) % 2 == 0
+    N = len(commitments) // 2
+    r = hash_to_field(commitments)
+    K = next_power_of_two(2 * N)
+    d = K - N - 1
+    r_to_K = pow(r, N, K)
+    roots = roots_of_unity(K)
+
+    # For an efficient implementation, B and Bprime should be precomputed
+    def B(z):
+        r = 1
+        for w in roots[:d + 1]:
+            r = r * (z - w) % MODULUS
+        return r
+
+    def Bprime(z):
+        r = 0
+        for i in range(d + 1):
+            m = 1
+            for w in roots[:i] + roots[i+1:d + 1]:
+                m = m * (z - w) % MODULUS
+            r = (r + M) % MODULUS
+        return r
+
+    coefs = []
+    for i in range(K):
+        coefs.append( - (r_to_K - 1) * modular_inverse(K * roots[i * (K - 1) % K] * (r - roots[i])) % MODULUS)
```

Comment:
Maybe to make it more understandable, make a variable `left_coeffs` and `right_coeffs`, and then at the `elliptic_curve_lincomb` line combine them together?